geode-protobuf/src/test/java/org/apache/geode/protocol/acceptance/CacheConnectionTimeoutJUnitTest.java

```diff
@@ -0,0 +1,161 @@
+
+  @Rule
+  public final RestoreSystemProperties restoreSystemProperties = new RestoreSystemProperties();
+
+  @Rule
+  public TestName testName = new TestName();
+
+  @Before
+  public void setup() throws Exception {
+    Properties properties = new Properties();
+
+    CacheFactory cacheFactory = new CacheFactory(properties);
+    cacheFactory.set(ConfigurationProperties.MCAST_PORT, "0");
+    cacheFactory.set(ConfigurationProperties.ENABLE_CLUSTER_CONFIGURATION, "false");
+    cacheFactory.set(ConfigurationProperties.USE_CLUSTER_CONFIGURATION, "false");
+
+    cache = cacheFactory.create();
+
+    CacheServer cacheServer = cache.addCacheServer();
+    cacheServerPort = AvailablePortHelper.getRandomAvailableTCPPort();
+    cacheServer.setPort(cacheServerPort);
+    cacheServer.setMaximumTimeBetweenPings(100);
+    cacheServer.start();
+
+    RegionFactory<Object, Object> regionFactory = cache.createRegionFactory();
+    regionFactory.create(TEST_REGION);
+
+    System.setProperty("geode.feature-protobuf-protocol", "true");
+    System.setProperty(ClientHealthMonitor.CLIENT_HEALTH_MONITOR_INTERVAL_PROPERTY, "100");
+
+    socket = new Socket("localhost", cacheServerPort);
+
+    Awaitility.await().atMost(5, TimeUnit.SECONDS).until(socket::isConnected);
+    outputStream = socket.getOutputStream();
+    outputStream.write(110);
+
+    serializationService = new ProtobufSerializationService();
+  }
+
+  @After
+  public void cleanup() throws IOException {
+    cache.close();
+    socket.close();
+    SocketCreatorFactory.close();
+  }
+
+  @Test
+  public void testUnresponsiveClientsGetDisconnected() throws Exception {
+    ProtobufProtocolSerializer protobufProtocolSerializer = new ProtobufProtocolSerializer();
+    ClientProtocol.Message putMessage =
+        MessageUtil.makePutRequestMessage(serializationService, TEST_KEY, TEST_VALUE, TEST_REGION,
+            ProtobufUtilities.createMessageHeader(TEST_PUT_CORRELATION_ID));
+
+    int pollInterval = 20;
+    int maximumTimeBetweenPings = ClientHealthMonitor.getInstance().getMaximumTimeBetweenPings();
+    long monitorInterval = ClientHealthMonitor.getInstance().getMonitorInterval();
+    long timeout = maximumTimeBetweenPings + monitorInterval + pollInterval;
+
+    // wait for client to get disconnected
+    Awaitility.await().atMost(timeout, TimeUnit.MILLISECONDS)
+        .pollInterval(pollInterval, TimeUnit.MILLISECONDS)
+        .pollDelay(maximumTimeBetweenPings + monitorInterval, TimeUnit.MILLISECONDS).until(() -> {
+          try {
+            /*
+             * send a PUT message
+             *
+             * Note: The `await` will run this at an interval larger than the maximum timeout
+             * allowed between pings
+             */
+            protobufProtocolSerializer.serialize(putMessage, outputStream);
+            assertEquals(-1, socket.getInputStream().read());
+          } catch (IOException e) {
+            e.printStackTrace();
+          }
+        });
+  }
```

Comment:
I'm having trouble getting my head around this check.  Why do you send a PUT to the server?  The health monitor will time out the connection created after the protobuf communication mode byte is sent.  Can't you just wait for a -1 w/o sending a message to the server?

Reply:
Yes, the test would still pass without that. However, I also want to ensure that the connection gets closed if the pings come at a rate close to the maximum allowed time between pings. The bug I am trying to catch is that we aren't correctly using the maximum time between pings, but some number `n | maximumTimeBetweenPings < n < timeout`